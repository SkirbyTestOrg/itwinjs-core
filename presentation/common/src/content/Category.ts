/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Content */

/**
 * A data structure that describes a [[Field]] category.
 * @public
 */
export interface CategoryDescription {
  /** Unique name */
  name: string;
  /** Display label */
  label: string;
  /** Extensive description */
  description: string;
  /** Priority. Categories with higher priority should appear higher in the UI */
  priority: number;
  /** Should this category be auto-expanded when it's displayed in the UI */
  expand: boolean;
}
