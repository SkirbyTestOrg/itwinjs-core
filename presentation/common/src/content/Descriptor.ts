/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Content */

import * as ec from "../EC";
import { Field, FieldJSON } from "./Fields";

/**
 * Data structure that describes an ECClass in content [[Descriptor]].
 */
export interface SelectClassInfo {
  /** Information about the ECClass */
  selectClassInfo: ec.ClassInfo;
  /** Is the class handled polymorphically */
  isSelectPolymorphic: boolean;
  /** Relationship path to the [Primary class]($docs/learning/content/Terminology#primary-class) */
  pathToPrimaryClass: ec.RelationshipPathInfo;
  /** Relationship paths to [Related property]($docs/learning/content/Terminology#related-properties) classes */
  relatedPropertyPaths: ec.RelationshipPathInfo[];
}

/**
 * Serialized [[SelectClassInfo]]
 *
 * @hidden
 */
export interface SelectClassInfoJSON {
  selectClassInfo: ec.ClassInfoJSON;
  isSelectPolymorphic: boolean;
  pathToPrimaryClass: ec.RelationshipPathInfoJSON;
  relatedPropertyPaths: ec.RelationshipPathInfoJSON[];
}

const selectClassInfoFromJSON = (json: SelectClassInfoJSON): SelectClassInfo => {
  return {
    ...json,
    selectClassInfo: ec.classInfoFromJSON(json.selectClassInfo),
    pathToPrimaryClass: json.pathToPrimaryClass.map((p) => ec.relatedClassInfoFromJSON(p)),
    relatedPropertyPaths: json.relatedPropertyPaths.map((rp) => (rp.map((p) => ec.relatedClassInfoFromJSON(p)))),
  };
};

/**
 * Flags that control content format.
 */
export enum ContentFlags {
  /** Each content record only has [[InstanceKey]] and no data */
  KeysOnly = 1 << 0,

  /** Each content record additionally has an image id */
  ShowImages = 1 << 1,

  /** Each content record additionally has a display label */
  ShowLabels = 1 << 2,

  /** All content records are merged into a single record (see [Merging values]($docs/learning/content/Terminology#value-merging)) */
  MergeResults = 1 << 3,

  /** Content has only distinct values */
  DistinctValues = 1 << 4,

  /** Doesn't create property or calculated fields. Can be used in conjunction with [[ShowLabels]]. */
  NoFields = 1 << 5,
}

/**
 * Data sorting direction
 */
export enum SortDirection {
  Ascending,
  Descending,
}

/**
 * Data structure that contains selection information. Used
 * for cases when requesting content after a selection change.
 */
export interface SelectionInfo {
  providerName: string;
  level?: number;
}

/**
 * Serialized [[Descriptor]] JSON representation.
 *
 * @hidden
 */
export interface DescriptorJSON {
  connectionId: string;
  inputKeysHash: string;
  contentOptions: any;
  selectionInfo?: SelectionInfo;
  displayType: string;
  selectClasses: SelectClassInfoJSON[];
  fields: FieldJSON[];
  sortingFieldName?: string;
  sortDirection?: SortDirection;
  contentFlags: number;
  filterExpression?: string;
}

/** @hidden */
export interface DescriptorOverrides {
  displayType: string;
  hiddenFieldNames: string[];
  contentFlags: number;
  sortingFieldName?: string;
  sortDirection?: SortDirection;
  filterExpression?: string;
}

/**
 * Data structure that describes content: fields, sorting, filtering, format, etc.
 * Descriptor may be changed to control how content is created.
 */
export default class Descriptor {
  /** Id of the connection used to create the descriptor */
  public connectionId!: string;
  /** Hash of the input keys used to create the descriptor */
  public inputKeysHash!: string;
  /** Extended options used to create the descriptor */
  public contentOptions: any;
  /** Selection info used to create the descriptor */
  public selectionInfo?: SelectionInfo;
  /** Display type used to create the descriptor */
  public displayType!: string;
  /** A list of classes that will be selected from when creating content with this descriptor */
  public selectClasses!: SelectClassInfo[];
  /** A list of fields contained in the descriptor */
  public fields!: Field[];
  /** [[ContentFlags]] used to create the descriptor */
  public contentFlags!: number;
  /** Field used to sort the content */
  public sortingField?: Field;
  /** Sorting direction */
  public sortDirection?: SortDirection;
  /** Content filtering [ECExpression]($docs/learning/ECExpressions) */
  public filterExpression?: string;

  /* istanbul ignore next */
  private constructor() { }

  /*public toJSON(): DescriptorJSON {
    return Object.assign({}, this, {
      fields: this.fields.map((field: Field) => field.toJSON()),
    });
  }*/

  /**
   * Deserialize Descriptor from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized descriptor or undefined if deserialization failed
   *
   * @hidden
   */
  public static fromJSON(json: DescriptorJSON | string | undefined): Descriptor | undefined {
    if (!json)
      return undefined;
    if (typeof json === "string")
      return JSON.parse(json, Descriptor.reviver);
    const descriptor = Object.create(Descriptor.prototype);
    return Object.assign(descriptor, json, {
      fields: json.fields.map((fieldJson: FieldJSON) => Field.fromJSON(fieldJson)),
      selectClasses: json.selectClasses.map((selectClass: SelectClassInfoJSON) => selectClassInfoFromJSON(selectClass)),
    } as Partial<Descriptor>);
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing Content objects.
   *
   * @hidden
   */
  public static reviver(key: string, value: any): any {
    return key === "" ? Descriptor.fromJSON(value) : value;
  }

  /**
   * Get field by its name
   * @param name Name of the field to find
   * @param recurse Recurse into nested fields
   */
  public getFieldByName(name: string, recurse?: boolean): Field | undefined {
    return findField(this.fields, name, recurse);
  }

  /** @hidden */
  public createDescriptorOverrides(): DescriptorOverrides {
    return {
      displayType: this.displayType,
      hiddenFieldNames: [],
      sortingFieldName: this.sortingField ? this.sortingField.name : undefined,
      sortDirection: this.sortDirection,
      contentFlags: this.contentFlags,
      filterExpression: this.filterExpression,
    };
  }

  /** @hidden */
  public resetParentship(): void {
    for (const field of this.fields)
      field.resetParentship();
  }

  /** @hidden */
  public rebuildParentship(): void {
    for (const field of this.fields)
      field.rebuildParentship();
  }

  /** @hidden */
  public createStrippedDescriptor(): Descriptor {
    const stripped = Object.create(Descriptor.prototype);
    return Object.assign(stripped, this, {
      fields: [],
      selectClasses: [],
    });
  }
}

const findField = (fields: Field[], name: string, recurse?: boolean): Field | undefined => {
  for (const field of fields) {
    if (field.name === name)
      return field;

    if (recurse && field.isNestedContentField()) {
      const nested = findField(field.nestedFields, name, recurse);
      if (nested)
        return nested;
    }
  }
  return undefined;
};
