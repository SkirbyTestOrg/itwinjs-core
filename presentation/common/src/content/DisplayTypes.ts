/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Content */

/**
 * Displayed content types. Affects how the content is formatted, e.g.
 * the [[ContentFlags]]
 */
export default class DefaultContentDisplayTypes {
  /** Unknown content type. */
  public static readonly UNDEFINED = "Undefined";

  /** Grid or table view content type. By default adds [[ContentFlags.ShowLabels]] flag. */
  public static readonly GRID = "Grid";

  /** Property pane content type. By default adds [[ContentFlags.MergeResults]] flag. */
  public static readonly PROPERTY_PANE = "PropertyPane";

  /** List content type. By default adds [[ContentFlags.NoFields]] and [[ContentFlags.ShowLabels]] flags */
  public static readonly LIST = "List";

  /** Viewport content type. By default adds [[ContentFlags.KeysOnly]] flag. */
  public static readonly VIEWPORT = "Graphics";
}
