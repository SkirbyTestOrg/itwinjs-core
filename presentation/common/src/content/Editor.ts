/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Content */

/**
 * A data structure which describes property editor used
 * for a content [[Field]].
 *
 * @public
 */
export interface EditorDescription {
  /** Unique name */
  name: string;
  /** Editor-specific parameters */
  params: any;
}
