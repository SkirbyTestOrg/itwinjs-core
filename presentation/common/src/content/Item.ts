/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @module Content */

import {
  ClassInfo, ClassInfoJSON,
  InstanceKey, InstanceKeyJSON,
} from "../EC";
import { ValuesDictionary } from "../Utils";
import {
  Value, DisplayValue,
  DisplayValueJSON, ValueJSON,
  ValuesMapJSON, DisplayValuesMapJSON,
} from "./Value";

/**
 * Serialized [[Item]] JSON representation.
 * @internal
 */
export interface ItemJSON {
  primaryKeys: InstanceKeyJSON[];
  label: string;
  imageId: string;
  classInfo?: ClassInfoJSON;
  values: ValuesDictionary<ValueJSON>;
  displayValues: ValuesDictionary<DisplayValueJSON>;
  mergedFieldNames: string[];
  extendedData?: { [key: string]: any };
}

/**
 * A data structure that represents a single content record.
 * @public
 */
export class Item {
  /** Keys of instances whose data is contained in this item */
  public primaryKeys: InstanceKey[];
  /** Display label of the item */
  public label: string;
  /** ID of the image associated with this item */
  public imageId: string;
  /** For cases when item consists only of same class instances, information about the ECClass */
  public classInfo?: ClassInfo;
  /** Raw values dictionary */
  public values: ValuesDictionary<Value>;
  /** Display values dictionary */
  public displayValues: ValuesDictionary<DisplayValue>;
  /** List of field names whose values are merged (see [Merging values]($docs/learning/presentation/Content/Terminology#value-merging)) */
  public mergedFieldNames: string[];
  /** Extended data injected into this content item */
  public extendedData?: { [key: string]: any };

  /**
   * Creates an instance of Item.
   * @param primaryKeys Keys of instances whose data is contained in this item
   * @param label Display label of the item
   * @param imageId ID of the image associated with this item
   * @param classInfo For cases when item consists only of same class instances, information about the ECClass
   * @param values Raw values dictionary
   * @param displayValues Display values dictionary
   * @param mergedFieldNames List of field names whose values are merged (see [Merging values]($docs/learning/presentation/Content/Terminology#value-merging))
   * @param extendedData Extended data injected into this content item
   */
  public constructor(primaryKeys: InstanceKey[], label: string, imageId: string, classInfo: ClassInfo | undefined,
    values: ValuesDictionary<Value>, displayValues: ValuesDictionary<DisplayValue>, mergedFieldNames: string[], extendedData?: { [key: string]: any }) {
    this.primaryKeys = primaryKeys;
    this.label = label;
    this.imageId = imageId;
    this.classInfo = classInfo;
    this.values = values;
    this.displayValues = displayValues;
    this.mergedFieldNames = mergedFieldNames;
    this.extendedData = extendedData;
  }

  /**
   * Is value of field with the specified name merged in this record.
   */
  public isFieldMerged(fieldName: string): boolean {
    return -1 !== this.mergedFieldNames.indexOf(fieldName);
  }

  /** @internal */
  public toJSON(): ItemJSON {
    return Object.assign({}, this, {
      classInfo: this.classInfo ? ClassInfo.toJSON(this.classInfo) : undefined,
      values: Value.toJSON(this.values) as ValuesMapJSON,
      displayValues: DisplayValue.toJSON(this.displayValues) as DisplayValuesMapJSON,
    });
  }

  /**
   * Deserialize Item from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized item or undefined if deserialization failed
   *
   * @internal
   */
  public static fromJSON(json: ItemJSON | string | undefined): Item | undefined {
    if (!json)
      return undefined;
    if (typeof json === "string")
      return JSON.parse(json, Item.reviver);
    const item = Object.create(Item.prototype);
    return Object.assign(item, json, {
      primaryKeys: json.primaryKeys.map((pk) => InstanceKey.fromJSON(pk)),
      classInfo: json.classInfo ? ClassInfo.fromJSON(json.classInfo) : undefined,
      values: Value.fromJSON(json.values),
      displayValues: DisplayValue.fromJSON(json.displayValues),
    } as Partial<Item>);
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing Item objects.
   *
   * @internal
   */
  public static reviver(key: string, value: any): any {
    return key === "" ? Item.fromJSON(value) : value;
  }
}
