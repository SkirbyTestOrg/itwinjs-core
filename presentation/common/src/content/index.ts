/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Content */

export { default as CategoryDescription } from "./Category";
export { default as Content } from "./Content";
export { default as Descriptor, SelectClassInfo, SelectionInfo, SortDirection, ContentFlags } from "./Descriptor";
export { default as DefaultContentDisplayTypes } from "./DisplayTypes";
export { default as EditorDescription } from "./Editor";
export { Field, PropertiesField, NestedContentField } from "./Fields";
export { default as Item } from "./Item";
export { default as Property } from "./Property";
export { PropertyValueFormat, TypeDescription, PrimitiveTypeDescription, ArrayTypeDescription, StructTypeDescription } from "./TypeDescription";
export {
  Value, ValuesArray, ValuesMap,
  DisplayValue, DisplayValuesArray, DisplayValuesMap,
  NestedContentValue,
} from "./Value";
