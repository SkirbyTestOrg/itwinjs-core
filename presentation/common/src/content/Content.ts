/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Content */

import Descriptor, { DescriptorJSON } from "./Descriptor";
import Item, { ItemJSON } from "./Item";

/**
 * Serialized [[Content]] JSON representation.
 *
 * @hidden
 */
export interface ContentJSON {
  descriptor: DescriptorJSON;
  contentSet: ItemJSON[];
}

/**
 * A data structure that contains the [[Descriptor]] and a list of [[Item]]
 * objects which are based on that descriptor.
 */
export default class Content {
  /** Descriptor used to create the content */
  public descriptor: Readonly<Descriptor>;
  /** Content items */
  public contentSet: Array<Readonly<Item>>;

  /* istanbul ignore next */
  private constructor(descriptor: Readonly<Descriptor>, items: Array<Readonly<Item>>) {
    this.descriptor = descriptor;
    this.contentSet = items;
  }

  /*public toJSON(): ContentJSON {
    return {
      descriptor: this.descriptor.toJSON(),
      contentSet: this.contentSet.map((item: Item) => item.toJSON()),
    };
  }*/

  /**
   * Deserialize Content from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized content or undefined if deserialization failed
   *
   * @hidden
   */
  public static fromJSON(json: ContentJSON | string | undefined): Content | undefined {
    if (!json)
      return undefined;
    if (typeof json === "string")
      return JSON.parse(json, Content.reviver);
    const content = Object.create(Content.prototype);
    return Object.assign(content, {
      descriptor: Descriptor.fromJSON(json.descriptor),
      contentSet: json.contentSet.map((itemJson: ItemJSON) => Item.fromJSON(itemJson)),
    });
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing Content objects.
   *
   * @hidden
   */
  public static reviver(key: string, value: any): any {
    return key === "" ? Content.fromJSON(value) : value;
  }
}
