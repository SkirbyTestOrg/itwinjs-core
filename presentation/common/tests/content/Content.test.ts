/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import * as faker from "faker";
import { createRandomECInstanceKeyJSON, createRandomECClassInfoJSON, createRandomDescriptorJson } from "@helpers/random";
import { Content } from "@src/content";
import { ContentJSON } from "@src/content/Content";

describe("Content", () => {

  describe("fromJSON", () => {

    let testContentJSON!: ContentJSON;
    beforeEach(() => {
      testContentJSON = {
        descriptor: createRandomDescriptorJson(),
        contentSet: [{
          classInfo: createRandomECClassInfoJSON(),
          primaryKeys: [createRandomECInstanceKeyJSON()],
          label: faker.random.words(),
          imageId: faker.random.uuid(),
          mergedFieldNames: [],
          values: {
            test: faker.random.number(),
          },
          displayValues: {
            test: faker.random.words(),
          },
        }],
      };
    });

    it("creates valid Content from valid JSON", () => {
      const item = Content.fromJSON(testContentJSON);
      expect(item).to.matchSnapshot();
    });

    it("creates valid Content from valid serialized JSON", () => {
      const item = Content.fromJSON(JSON.stringify(testContentJSON));
      expect(item).to.matchSnapshot();
    });

    it("returns undefined for undefined JSON", () => {
      const item = Content.fromJSON(undefined);
      expect(item).to.be.undefined;
    });

  });

});
