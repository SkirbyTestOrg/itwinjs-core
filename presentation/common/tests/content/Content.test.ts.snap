// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content fromJSON creates valid Content from valid JSON 1`] = `
Content {
  "contentSet": Array [
    Item {
      "classInfo": Object {
        "id": "0x4c8b00000131af",
        "label": "deposit",
        "name": "Rustic Plastic Cheese",
      },
      "displayValues": Object {
        "test": "payment",
      },
      "imageId": "d4edf155-65b5-47fc-8d51-dcdf22654a29",
      "label": "Generic Frozen Hat Front-line COM",
      "mergedFieldNames": Array [],
      "primaryKeys": Array [
        Object {
          "className": "parsing",
          "id": "0x8ae600000070d1",
        },
      ],
      "values": Object {
        "test": 62323,
      },
    },
  ],
  "descriptor": Descriptor {
    "connectionId": "ae351acd-eef3-43c8-995b-67c79b260315",
    "contentFlags": 0,
    "contentOptions": "bar",
    "displayType": "qui voluptatem repellendus",
    "fields": Array [
      Field {
        "category": Object {
          "description": "Molestiae maxime culpa aliquam.",
          "expand": true,
          "label": "mobile",
          "name": "Handcrafted",
          "priority": 52751,
        },
        "editor": undefined,
        "isReadonly": false,
        "label": "Baby capacitor utilize",
        "name": "payment",
        "priority": 29468,
        "type": Object {
          "typeName": "tinyint",
          "valueFormat": "Primitive",
        },
      },
      Field {
        "category": Object {
          "description": "Rerum amet odio libero tenetur numquam.",
          "expand": false,
          "label": "SMTP",
          "name": "firmware",
          "priority": 23381,
        },
        "editor": undefined,
        "isReadonly": false,
        "label": "homogeneous",
        "name": "program",
        "priority": 66562,
        "type": Object {
          "typeName": "float",
          "valueFormat": "Primitive",
        },
      },
      Field {
        "category": Object {
          "description": "Qui aut cum nam voluptas dolorem.",
          "expand": false,
          "label": "Distributed Fantastic",
          "name": "collaboration",
          "priority": 39754,
        },
        "editor": Object {
          "name": "quantify",
        },
        "isReadonly": true,
        "label": "Computer Roads utilisation",
        "name": "primary",
        "priority": 48282,
        "type": Object {
          "typeName": "varchar",
          "valueFormat": "Primitive",
        },
      },
    ],
    "inputKeysHash": "396a8648-6f9e-4185-bb78-a43adbe96ce2",
    "selectClasses": Array [
      Object {
        "isSelectPolymorphic": true,
        "pathToPrimaryClass": Array [
          Object {
            "isForwardRelationship": false,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0x157f4000000bd90",
              "label": "Markets bypassing Arizona",
              "name": "Kids",
            },
            "sourceClassInfo": Object {
              "id": "0x16bab0000006ce3",
              "label": "Avon Baby",
              "name": "payment",
            },
            "targetClassInfo": Object {
              "id": "0xde29000000d857",
              "label": "Mews Tonga multi-state",
              "name": "Refined Frozen Gloves",
            },
          },
          Object {
            "isForwardRelationship": true,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0xb60a000000bd54",
              "label": "matrix",
              "name": "deposit",
            },
            "sourceClassInfo": Object {
              "id": "0x275400000130f4",
              "label": "Sports Steel program",
              "name": "AI",
            },
            "targetClassInfo": Object {
              "id": "0x52f20000007465",
              "label": "Berkshire",
              "name": "capability",
            },
          },
        ],
        "relatedPropertyPaths": Array [
          Array [
            Object {
              "isForwardRelationship": true,
              "isPolymorphicRelationship": true,
              "relationshipInfo": Object {
                "id": "0xb639000000433c",
                "label": "withdrawal Soap",
                "name": "Sleek Frozen Pants",
              },
              "sourceClassInfo": Object {
                "id": "0x18698000000584e",
                "label": "Missouri bluetooth",
                "name": "Consultant",
              },
              "targetClassInfo": Object {
                "id": "0x11d1a00000151c2",
                "label": "olive withdrawal Nevada",
                "name": "Computers",
              },
            },
          ],
          Array [
            Object {
              "isForwardRelationship": false,
              "isPolymorphicRelationship": false,
              "relationshipInfo": Object {
                "id": "0x247f00000000ed",
                "label": "Tactics Refined",
                "name": "Enhanced",
              },
              "sourceClassInfo": Object {
                "id": "0x132e5000000e6e3",
                "label": "Haiti",
                "name": "sexy",
              },
              "targetClassInfo": Object {
                "id": "0xa0c400000179f4",
                "label": "index mint green open system",
                "name": "exploit",
              },
            },
          ],
        ],
        "selectClassInfo": Object {
          "id": "0x8634000001605b",
          "label": "Associate",
          "name": "Ergonomic",
        },
      },
      Object {
        "isSelectPolymorphic": true,
        "pathToPrimaryClass": Array [
          Object {
            "isForwardRelationship": false,
            "isPolymorphicRelationship": false,
            "relationshipInfo": Object {
              "id": "0xb884000000c643",
              "label": "violet Marshall Islands",
              "name": "JSON",
            },
            "sourceClassInfo": Object {
              "id": "0x22c0000000ed76",
              "label": "Tunnel",
              "name": "Intelligent",
            },
            "targetClassInfo": Object {
              "id": "0x183a60000014ffb",
              "label": "Chief Virtual",
              "name": "Outdoors",
            },
          },
          Object {
            "isForwardRelationship": false,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0x141010000006eea",
              "label": "application Soft black",
              "name": "Austria",
            },
            "sourceClassInfo": Object {
              "id": "0x17c180000003961",
              "label": "incentivize Bike Supervisor",
              "name": "Security",
            },
            "targetClassInfo": Object {
              "id": "0x117b800000132c3",
              "label": "Group copying",
              "name": "Libyan Dinar",
            },
          },
        ],
        "relatedPropertyPaths": Array [
          Array [
            Object {
              "isForwardRelationship": true,
              "isPolymorphicRelationship": true,
              "relationshipInfo": Object {
                "id": "0xf9a40000014e2d",
                "label": "Mobility Personal Loan Account",
                "name": "collaboration",
              },
              "sourceClassInfo": Object {
                "id": "0x4f400000182d6",
                "label": "Officer",
                "name": "withdrawal",
              },
              "targetClassInfo": Object {
                "id": "0x10315000000a8e2",
                "label": "Specialist Forks",
                "name": "Focused",
              },
            },
          ],
          Array [
            Object {
              "isForwardRelationship": true,
              "isPolymorphicRelationship": false,
              "relationshipInfo": Object {
                "id": "0xd87000000bc51",
                "label": "mint green Minnesota",
                "name": "program",
              },
              "sourceClassInfo": Object {
                "id": "0x15fd90000005b5c",
                "label": "El Salvador New York",
                "name": "Guinea-Bissau",
              },
              "targetClassInfo": Object {
                "id": "0x9fa00000015c2e",
                "label": "feed Neck",
                "name": "deposit",
              },
            },
          ],
        ],
        "selectClassInfo": Object {
          "id": "0x166d200000180ff",
          "label": "Producer",
          "name": "well-modulated",
        },
      },
    ],
  },
}
`;

exports[`Content fromJSON creates valid Content from valid serialized JSON 1`] = `
Content {
  "contentSet": Array [
    Item {
      "classInfo": Object {
        "id": "0x1821b000001150a",
        "label": "synthesizing encryption",
        "name": "Pennsylvania",
      },
      "displayValues": Object {
        "test": "Intelligent Cotton Pants",
      },
      "imageId": "cd3956f0-152c-47ef-9f41-8fe6e8dd70b4",
      "label": "Macao coherent Investment Account",
      "mergedFieldNames": Array [],
      "primaryKeys": Array [
        Object {
          "className": "Orchestrator",
          "id": "0xadaf0000006152",
        },
      ],
      "values": Object {
        "test": 88875,
      },
    },
  ],
  "descriptor": Descriptor {
    "connectionId": "c7cc673b-0110-4180-8565-cc9aff73bb20",
    "contentFlags": 0,
    "contentOptions": "bar",
    "displayType": "aut nostrum repudiandae",
    "fields": Array [
      Field {
        "category": Object {
          "description": "Ea et nisi velit nam ab enim eligendi.",
          "expand": false,
          "label": "Skyway",
          "name": "Berkshire",
          "priority": 7517,
        },
        "editor": Object {
          "name": "Steel",
        },
        "isReadonly": false,
        "label": "Money Market Account Soft open-source",
        "name": "integrated",
        "priority": 35016,
        "type": Object {
          "typeName": "datetime",
          "valueFormat": "Primitive",
        },
      },
      Field {
        "category": Object {
          "description": "Nulla perferendis provident quasi dolores doloribus voluptatem ut.",
          "expand": true,
          "label": "Hat Dynamic Avon",
          "name": "multi-state",
          "priority": 14220,
        },
        "isReadonly": false,
        "label": "1080p streamline CSS",
        "name": "Fish",
        "priority": 46357,
        "type": Object {
          "typeName": "bit",
          "valueFormat": "Primitive",
        },
      },
      Field {
        "category": Object {
          "description": "Id aut veniam eum dolorum qui.",
          "expand": false,
          "label": "Toys cyan",
          "name": "Program",
          "priority": 86529,
        },
        "isReadonly": true,
        "label": "Rand Loti HDD Car",
        "name": "unleash",
        "priority": 16225,
        "type": Object {
          "typeName": "date",
          "valueFormat": "Primitive",
        },
      },
    ],
    "inputKeysHash": "a3856a41-05be-4558-a808-72be4376380f",
    "selectClasses": Array [
      Object {
        "isSelectPolymorphic": false,
        "pathToPrimaryClass": Array [
          Object {
            "isForwardRelationship": true,
            "isPolymorphicRelationship": false,
            "relationshipInfo": Object {
              "id": "0x13ce900000031a5",
              "label": "Green",
              "name": "Shirt",
            },
            "sourceClassInfo": Object {
              "id": "0xef810000005e49",
              "label": "Brand",
              "name": "aggregate",
            },
            "targetClassInfo": Object {
              "id": "0x9a390000009a43",
              "label": "wireless",
              "name": "Mouse",
            },
          },
          Object {
            "isForwardRelationship": false,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0xaf230000007dcc",
              "label": "Fish Investment Account withdrawal",
              "name": "Assistant",
            },
            "sourceClassInfo": Object {
              "id": "0xd0930000006b05",
              "label": "Avon Generic",
              "name": "Marketing",
            },
            "targetClassInfo": Object {
              "id": "0x58ce000000257a",
              "label": "Borders",
              "name": "Soft",
            },
          },
        ],
        "relatedPropertyPaths": Array [
          Array [
            Object {
              "isForwardRelationship": true,
              "isPolymorphicRelationship": false,
              "relationshipInfo": Object {
                "id": "0x151020000017f67",
                "label": "Music",
                "name": "invoice",
              },
              "sourceClassInfo": Object {
                "id": "0x897f0000000822",
                "label": "e-tailers",
                "name": "data-warehouse",
              },
              "targetClassInfo": Object {
                "id": "0x61140000001644",
                "label": "tangible",
                "name": "deposit",
              },
            },
          ],
          Array [
            Object {
              "isForwardRelationship": true,
              "isPolymorphicRelationship": false,
              "relationshipInfo": Object {
                "id": "0x183f10000016c8f",
                "label": "Kazakhstan",
                "name": "Music",
              },
              "sourceClassInfo": Object {
                "id": "0x1401400000005d4",
                "label": "Cambridgeshire Uruguay Netherlands Antilles",
                "name": "Supervisor",
              },
              "targetClassInfo": Object {
                "id": "0x2227000000693d",
                "label": "Chair",
                "name": "Colombia",
              },
            },
          ],
        ],
        "selectClassInfo": Object {
          "id": "0xc9b1000000e4ef",
          "label": "Intuitive",
          "name": "Handmade Metal Soap",
        },
      },
      Object {
        "isSelectPolymorphic": false,
        "pathToPrimaryClass": Array [
          Object {
            "isForwardRelationship": false,
            "isPolymorphicRelationship": true,
            "relationshipInfo": Object {
              "id": "0x735c0000006cc4",
              "label": "Devolved navigate payment",
              "name": "Tasty Fresh Chair",
            },
            "sourceClassInfo": Object {
              "id": "0x2b0b00000112fa",
              "label": "Regional",
              "name": "Metal",
            },
            "targetClassInfo": Object {
              "id": "0x1018e000000a96d",
              "label": "North Carolina",
              "name": "protocol",
            },
          },
          Object {
            "isForwardRelationship": false,
            "isPolymorphicRelationship": false,
            "relationshipInfo": Object {
              "id": "0x17e3f0000011cf3",
              "label": "solutions Guarani",
              "name": "French Southern Territories",
            },
            "sourceClassInfo": Object {
              "id": "0x165140000006eb6",
              "label": "Avon embrace",
              "name": "hard drive",
            },
            "targetClassInfo": Object {
              "id": "0x49090000002772",
              "label": "implementation Steel navigate",
              "name": "Auto Loan Account",
            },
          },
        ],
        "relatedPropertyPaths": Array [
          Array [
            Object {
              "isForwardRelationship": true,
              "isPolymorphicRelationship": true,
              "relationshipInfo": Object {
                "id": "0x403e000001232d",
                "label": "systemic",
                "name": "Soft",
              },
              "sourceClassInfo": Object {
                "id": "0xb5c100000050d1",
                "label": "Marketing pink e-markets",
                "name": "Business-focused",
              },
              "targetClassInfo": Object {
                "id": "0x151e40000000860",
                "label": "Fresh Lead Factors",
                "name": "leverage",
              },
            },
          ],
          Array [
            Object {
              "isForwardRelationship": false,
              "isPolymorphicRelationship": false,
              "relationshipInfo": Object {
                "id": "0x15293000000b747",
                "label": "orange Somoni Small Plastic Chicken",
                "name": "deliver",
              },
              "sourceClassInfo": Object {
                "id": "0x54e9000000fc5d",
                "label": "eco-centric",
                "name": "Networked",
              },
              "targetClassInfo": Object {
                "id": "0xdd43000001291c",
                "label": "invoice synthesizing leading-edge",
                "name": "Outdoors",
              },
            },
          ],
        ],
        "selectClassInfo": Object {
          "id": "0xa30c000000d87b",
          "label": "Analyst Mozambique Burundi Franc",
          "name": "Assistant",
        },
      },
    ],
  },
}
`;
